### DISCUSSION TITLE
How to share package dependencies

### DISCUSSION THREAD

**Discussion by craigcomstock** [cfengine/core#4512] [Q&A]
I want to ensure that both ansible and cf-remote pip modules are installed and up-to-date but both rely on pip being installed.

I have this currently for ansible and was working on adding cfremote as a separate policy and then didn't quite know how to combine/share the pip package dependency.

```cf3
bundle agent ansible
{
  meta:
    role_dev.linux::
      "tags" slist => { "autorun" };

  vars:
    debian::
      "packages" slist => { "python3-pip" };
    alpine::
      "packages" slist => { "py3-pip" };

  packages:
    "$(packages)"
      policy => "present",
      handle => "pip_installed",
      version => "latest";
    "ansible"
      package_policy => "add",
      depends_on => { "pip_installed" },
      package_method => pip(""),
      handle => "ansible_installed";
}
```

and for cf-remote I would have much the same but I hesitate to copy/paste/modify! :O How best to share the "pip_installed" situation/dependency?

Ideally I'd like to be able to just share how to install cfremote with other folks via a "CFEngine policy package manager". Something like npm, say `cpm` and I do `cpm install cfremote` which has a dependency on pip so it adds policy that understands how to install and configure cf-remote as well as pip on various systems.

I would like the policy to be clear as well, stating the dependency on pip in both ansible and cfremote policy.

So I will go that way for now.

Any opinions? Existing strategies that you employ currently you want to share?

**Comment by nickanderson** [comment]
I think that the traditional answer to this would be that each policy should stand on its own. Following that, you would have the promises that pip is a required package as part of both policies.

Thinking up the abstraction layer, we would need some kind of tooling that facilitated sharing and integrating other people's policies. Then the policy could advertize a dependency on another policy. Unfortunately, there is no currently maintained tooling that does such a thing. There have been attempts in the past, Design Center comes to mind, but it did not thrive.
